### PR TITLE
Release v0.14.0 — Alpine-only image, integrity checking, verbose logging

### LINKED ISSUE #125: Add post-download file validation (integrity check)
## Summary
Add checksum-based integrity verification after file downloads to detect corrupt transfers.

## Details
RapidCopy implements chunk-based validation with adaptive sizing and auto-retry. Even a simpler whole-file checksum would cover most use cases.

## Implementation Options
- **Simple**: Whole-file hash comparison (remote vs local) after download. Algorithms: md5, sha256.
- **Advanced**: Chunk-based validation with configurable chunk size, adaptive sizing, inline validation during download.

## New File States
- VALIDATING (in progress)
- VALIDATED (passed)
- CORRUPT (failed — trigger re-download)

### Acceptance Criteria
- [ ] Configurable validation enable/disable and algorithm
- [ ] Per-file validation trigger (manual and auto)
- [ ] VALIDATING/VALIDATED/CORRUPT states in UI
- [ ] Corrupt files can be re-queued for download

### LINKED ISSUE #156: feat: File list pair label column
## Description

Once multiple path pairs produce files, users need to see which pair a file belongs to. Add a pair name/label column to the file list table, with optional filter/group by pair.

## Context

Deferred from PR #149. Depends on per-pair controller wiring being complete.

### LINKED ISSUE #225: Comprehensive test coverage improvements
## Overview

Test coverage audit identified significant gaps in both Angular frontend and Python backend testing. This issue tracks the remaining work after the v0.13.2 items (path pairs handler integration tests, file-list/bulk-action-bar component tests) are merged.

**Current state**: 182 Angular unit tests, ~30/60 Python source files tested. Key UX-critical paths lack coverage.

---

### 1. `config.service.ts` — Settings save/load flow
**Why it matters**: Users lose settings if this breaks silently.
**What to test**:
- Loading config from REST endpoint on init
- Saving config changes via PUT
- Error handling when backend is unreachable
- Config change notification to subscribers
**Files**: `src/angular/src/app/services/settings/config.service.ts`
**Prompt**:
> Read `config.service.ts` and `config.service.spec.ts` (if exists) and the existing test patterns in `settings-page.component.spec.ts`. Create `config.service.spec.ts` testing: initial config load from REST, config update via PUT, error handling on network failure, BehaviorSubject emission on config change. Mock `RestService` using the patterns from `model-file.service.spec.ts`.

### 2. `rest.service.ts` — HTTP request wrapper
**Why it matters**: Every API call flows through this service.
**What to test**:
- GET/POST/PUT/DELETE methods construct correct URLs
- Error responses are properly propagated
- Base URL configuration
**Files**: `src/angular/src/app/services/utils/rest.service.ts`
**Prompt**:
> Read `rest.service.ts`. Create `rest.service.spec.ts` testing all HTTP methods (GET, POST, PUT, DELETE) with `provideHttpClientTesting`. Verify URL construction, request body serialization, error propagation for 4xx/5xx responses, and timeout handling.

### 3. `stream-dispatch.service.ts` — WebSocket/SSE event dispatch
**Why it matters**: Real-time file status updates depend on this.
**What to test**:
- Event registration and callback dispatch
- Reconnection behavior on disconnect
- Multiple subscriber support
- Event type routing (model-init, model-added, model-updated, model-removed)
**Files**: `src/angular/src/app/services/utils/stream-dispatch.service.ts`
**Prompt**:
> Read `stream-dispatch.service.ts` and how `model-file.service.spec.ts` mocks it. Create `stream-dispatch.service.spec.ts` testing: event listener registration, event dispatch to correct callbacks, multiple listeners for same event type, listener removal, reconnection behavior after disconnect.

### 4. `api-key.interceptor.ts` — API key injection
**What to test**: Key is added to all outbound requests, missing key handling.
**Files**: `src/angular/src/app/services/utils/api-key.interceptor.ts`

---

### 5. `logs-page.component` — Log display and filtering
**What to test**:
- Log entries rendered from LogService
- Search/filter input updates displayed logs
- Level filter (INFO/WARNING/ERROR) works
- Scroll behavior / auto-scroll to latest
**Files**: `src/angular/src/app/pages/logs/logs-page.component.ts` + `.html`
**Prompt**:
> Read `logs-page.component.ts`, `.html`, and `log.service.ts`. Create `logs-page.component.spec.ts` testing: log entries render from service, search input filters displayed logs, level dropdown filters by severity, new log entries append correctly. Mock LogService with a Subject that emits test LogRecord objects.

### 6. `autoqueue-page.component` — Pattern management UI
**What to test**:
- Pattern list rendering
- Add pattern (valid input, empty input rejection)
- Remove pattern with confirmation
- Connected/disconnected state display
**Files**: `src/angular/src/app/pages/autoqueue/autoqueue-page.component.ts` + `.html`

### 7. `header.component` — Connection status and theme
**What to test**:
- Connection status indicator reflects ServerStatusService state
- Theme toggle calls ThemeService
- Version display
**Files**: `src/angular/src/app/pages/main/header.component.ts` + `.html`

### 8. `sidebar.component` — Navigation
**What to test**: Active route highlighting, navigation links.
**Files**: `src/angular/src/app/pages/main/sidebar.component.ts` + `.html`

### 9. `view-file-filter.service.ts` and `view-file-sort.service.ts`
**What to test**: Filter predicates, sort comparators, edge cases (empty list, single item).
**Prompt**:
> Read `view-file-filter.service.ts` and `view-file-sort.service.ts`. Create spec files testing all filter predicates (by status, by name substring, by pair) and all sort comparators (by name, size, status, progress). Test edge cases: empty array, single element, equal values.

---

### 10. `controller.py` — Main controller state machine
**Why it matters**: Core orchestration logic — file state transitions, command routing, pair context management.
**What to test**:
- `_persist_key` / `_strip_persist_key` round-trip
- `filter_excluded_files` (now has tests from #158 PR)
- Command dispatch (queue, stop, extract, delete) routes to correct pair context
- State machine: QUEUED → DOWNLOADING → DOWNLOADED → EXTRACTING → EXTRACTED
- Error state handling: EXTRACT_FAILED recovery
**Files**: `src/python/controller/controller.py`
**Prompt**:
> Read `controller.py` thoroughly. The Controller class is complex with per-pair contexts (_PairContext). Create unit tests in `src/python/tests/unittests/test_controller/test_controller_unit.py` (separate from integration tests). Test: `_persist_key`/`_strip_persist_key` round-trip with various inputs including filenames with special chars, command dispatch routing to correct pair context, state machine transitions via `_process_model_diffs`. Mock LFTP, scanner, and web dependencies.

### 11. `notifier.py` — Webhook notifications
**Status**: 7 tests added in PR #222 (shutdown drain). Still needs:
- Payload format verification
- URL validation (http/https scheme check)
- Timeout behavior on slow endpoints
- Event type filtering
**Files**: `src/python/controller/notifier.py`

### 12. `logs.py` web handler — Log file reading with search/filter
**What to test**:
- GET /server/logs returns log entries
- Search parameter filters by content
- Level parameter filters by severity
- Pagination (limit/offset)
- Empty log file handling
- Log file rotation handling
**Files**: `src/python/web/handler/logs.py`
**Prompt**:
> Read `src/python/web/handler/logs.py` and the test patterns in `src/python/tests/integration/test_web/test_config.py`. Create `src/python/tests/integration/test_web/test_logs.py` testing all query parameters: search, level filter, limit, offset. Test empty logs, large log files, and log rotation scenarios.

### 13. `security.py` — API key validation middleware
**What to test**: Valid key passes, invalid key returns 401, missing key returns 401, key rotation.
**Files**: `src/python/web/security.py`

---

### 14. `local_scanner.py` and `remote_scanner.py`
**What to test**: Directory walking, file discovery, symlink handling, permission errors.
**Files**: `src/python/controller/scan/local_scanner.py`, `src/python/controller/scan/remote_scanner.py`

### 15. `delete_process.py` — File deletion execution
**What to test**: Local delete, remote delete via SSH, error handling on permission denied.
**Files**: `src/python/controller/delete/delete_process.py`

---

### 16. Replace Protractor with Playwright
The current `src/e2e/` suite uses Protractor (deprecated since Angular 12, unmaintained). It requires Selenium and has ancient dependencies (TypeScript 3.9, `@types/node` v14).

**What to do**:
- Remove `src/e2e/` directory
- Create new Playwright test suite
- Add to CI pipeline (run against Docker container)
- Cover critical user flows:
  - App loads and shows file list
  - Settings page: add/edit/delete path pair
  - File operations: queue, stop, delete
  - Auto-queue: add/remove patterns
  - Logs page: view and filter logs
  - About page: version display

**Prompt**:
> Remove the `src/e2e/` directory (Protractor, deprecated). Create a new Playwright test suite at `e2e/` (root level). Initialize with `npm init playwright@latest`. Create page objects for: FilesPage, SettingsPage, AutoQueuePage, LogsPage, AboutPage. Write tests for critical user flows: app loads with file list visible, add a path pair in settings and verify it appears, queue/stop a file, add/remove auto-queue pattern, filter logs by level, about page shows version. Configure `playwright.config.ts` to run against `http://localhost:8800` (the Docker container). Add a CI job to `.github/workflows/ci.yml` that starts the container and runs Playwright tests.

---

## Priority Order

1. **v0.13.2** (in progress): Path pairs handler integration tests, file-list/bulk-action-bar component tests
2. **v0.14.0 Phase 1**: Items 1-3 (config.service, rest.service, stream-dispatch.service) — core communication layer
3. **v0.14.0 Phase 2**: Items 10, 12 (controller unit tests, logs handler) — backend gaps
4. **v0.14.0 Phase 3**: Items 5-9 (remaining component tests) — UI coverage
5. **v0.14.0 Phase 4**: Item 16 (Playwright e2e) — end-to-end confidence
6. **Backlog**: Items 11, 13-15 (notifier, security, scanners, delete)

### LINKED ISSUE #228: Switch from fork to spawn for multiprocessing to fix Python 3.12 deprecation warnings
## Problem

Python 3.12 emits deprecation warnings when `fork()` is called in a multi-threaded process:

```
/usr/local/lib/python3.12/multiprocessing/popen_fork.py:66: DeprecationWarning: This process (pid=1) is multi-threaded, use of fork() may lead to deadlocks in the child.
  self.pid = os.fork()
/usr/local/lib/python3.12/pty.py:95: DeprecationWarning: This process (pid=41) is multi-threaded, use of forkpty() may lead to deadlocks in the child.
  pid, fd = os.forkpty()
```

This happens because SeedSync starts child processes (`ActiveScanner`, `LocalScanner`, `RemoteScanner`, `ExtractProcess`, `DeleteLocalProcess`, `DeleteRemoteProcess`) **after** threads are already running (`ControllerJob`, `WebAppJob`, `MPLoggerListener`). Python's default `fork` start method copies all memory (including locked mutexes from other threads) into the child, which can cause deadlocks.

**Python 3.14 will change the default start method to `spawn` on Linux**, so this must be fixed before upgrading.

### Files that create subprocesses

| File | Process | Type |
|------|---------|------|
| `src/python/common/app_process.py` | `AppProcess(Process)` | Base class for all subprocesses |
| `src/python/controller/controller.py` | `ScannerProcess` x3, `ExtractProcess`, `DeleteLocalProcess`, `DeleteRemoteProcess` | Long-running + one-shot |
| `src/python/controller/extract/extract_process.py` | `ExtractProcess` | Long-running with worker thread |
| `src/python/controller/delete/delete_process.py` | `DeleteLocalProcess`, `DeleteRemoteProcess` | One-shot |

### Files that use pexpect (forkpty)

| File | Usage |
|------|-------|
| `src/python/lftp/lftp.py` | `pexpect.spawn("/usr/bin/lftp", ...)` |
| `src/python/ssh/sshcp.py` | `pexpect.spawn(command)` for SSH operations |

### Cross-process communication

| Mechanism | Location |
|-----------|----------|
| `multiprocessing.Queue` | `MultiprocessingLogger`, `ExtractProcess`, `AppProcess.__exception_queue` |
| `multiprocessing.Event` | `AppProcess._terminate` |

### 1. Set multiprocessing start method to `spawn`

In `seedsync.py`, before any subprocesses are created:

```python
import multiprocessing
multiprocessing.set_start_method('spawn')
```

### 2. Ensure all subprocess arguments are picklable

With `spawn`, child processes don't inherit parent memory — everything passed to the child must be picklable. Audit:

- `AppProcess.__init__` — `name: str` ✅
- `AppProcess.set_multiprocessing_logger` — `MultiprocessingLogger` contains a `multiprocessing.Queue` (picklable) but also a `threading.Thread` and `threading.Event` (NOT picklable). **Must refactor** to pass only the queue to child processes, not the whole `MultiprocessingLogger` object.
- `ScannerProcess` — receives config values, paths, queues. Should be picklable.
- `ExtractProcess` — receives `multiprocessing.Queue` instances. ✅
- `DeleteLocalProcess` / `DeleteRemoteProcess` — receive path strings. ✅

### 3. Refactor `MultiprocessingLogger` for spawn compatibility

Current design passes the entire `MultiprocessingLogger` object to child processes via `set_multiprocessing_logger()`. With `spawn`, this fails because `threading.Thread` and `threading.Event` aren't picklable.

**Fix**: Pass only the `multiprocessing.Queue` to child processes. The child creates its own `QueueHandler` from the queue. The listener thread stays in the parent.

```python
# In AppProcess
def set_multiprocessing_logger_queue(self, log_queue: multiprocessing.Queue, log_level: int):
    self._log_queue = log_queue
    self._log_level = log_level

# In AppProcess.run() — called in child process after spawn
def run(self):
    if hasattr(self, '_log_queue'):
        queue_handler = QueueHandler(self._log_queue)
        root = logging.getLogger()
        root.handlers.clear()
        root.addHandler(queue_handler)
        root.setLevel(self._log_level)
        self.logger = root.getChild(self.__name)
    ...
```

### 4. Handle pexpect `fork()` warnings

`pexpect.spawn()` uses `forkpty()` internally. This is harder to avoid since pexpect requires a PTY. Options:
- **Accept the warning** for pexpect — it forks inside an already-spawned child process (e.g., `RemoteScanner`), where there's only one thread, so it's safe.
- Verify that pexpect calls only happen inside child processes (not the multi-threaded parent). Currently:
  - `lftp.py` — spawned from `ControllerJob` thread in the **parent** process. **This is the problem.**
  - `sshcp.py` — used by `RemoteScanner` which runs in a **child** process. ✅ Safe.

**Fix for lftp.py**: The LFTP process is spawned from the parent's `ControllerJob` thread. Since LFTP is a long-running interactive process managed via pexpect, consider:
- Moving LFTP management into its own subprocess (cleanest)
- Or suppressing the warning since the forkpty happens from a known-safe state

### 5. Update signal handling in `AppProcess.run()`

The comment in `app_process.py:59-69` explains a race condition with signal handlers and `fork()`. With `spawn`, child processes start fresh (no inherited signal handlers), so this code can be simplified — the signal resets become unnecessary.

### 6. Remove fork-specific logger cleanup

In `multiprocessing_logger.py:61-66`, the comment says "The fork may have happened after the root logger was setup by the main process." With `spawn`, the child starts with a clean logger state, so this cleanup is no longer needed (but keeping it is harmless).

## Testing

- Verify no `DeprecationWarning` about `fork()` in logs
- Verify all scanner processes start and scan correctly
- Verify extract process works (queue a download with an archive)
- Verify delete operations work (both local and remote)
- Verify logging from child processes appears correctly in parent
- Verify LFTP downloads work end-to-end
- Run existing integration tests in Docker (both alpine and debian)

## Claude Code Prompt

```
Switch SeedSync from fork to spawn multiprocessing start method to fix Python 3.12 deprecation warnings.

Key changes needed:
1. Add `multiprocessing.set_start_method('spawn')` early in seedsync.py
2. Refactor MultiprocessingLogger — pass only the multiprocessing.Queue to child processes, not the full object (threading.Thread/Event aren't picklable)
3. Update AppProcess to configure logging from a queue in run() instead of inheriting the full MultiprocessingLogger
4. Simplify signal handling in AppProcess.run() — spawn doesn't inherit parent signal handlers
5. Address pexpect.spawn() in lftp.py which calls forkpty() from the multi-threaded parent process
6. Verify all subprocess constructor arguments are picklable
7. Run full integration tests in Docker to verify scanners, extraction, deletion, and LFTP all work

Files to modify:
- src/python/seedsync.py — set start method
- src/python/common/app_process.py — refactor logger setup, simplify signals
- src/python/common/multiprocessing_logger.py — expose queue for child processes
- src/python/controller/controller.py — update set_multiprocessing_logger calls
- src/python/controller/extract/extract_process.py — verify picklability
- src/python/lftp/lftp.py — address forkpty in threaded parent

Reference the deprecation warnings in the logs and test that they're gone after the change.
```

## References

- [Python 3.12 deprecation notice](https://docs.python.org/3.12/library/multiprocessing.html#contexts-and-start-methods)
- [Python 3.14 will default to spawn](https://github.com/python/cpython/issues/84559)
- Related: `app_process.py:59-69` signal handler race condition comment

### LINKED ISSUE #231: Remove Debian Docker variant — Alpine-only release
## Summary

Drop the Debian Docker image variant and ship only Alpine going forward. Optimize the remaining build pipeline for speed and image size.

## Rationale

- Alpine image is ~45 MB vs Debian at ~126 MB
- Maintaining two variants doubles CI build/publish time
- No known users depending on Debian-specific packages

### Debian Removal
- Remove Debian Dockerfile (`Dockerfile`) and keep only `Dockerfile.alpine`
- Update CI workflow (`ci.yml`) to remove Debian matrix entries
- Update `docker-compose.yml` and any docs referencing the Debian tag
- Update README / website with note about Alpine-only
- Ensure all runtime dependencies (lftp, 7zip, p7zip-rar, etc.) are available in Alpine

### CI Optimizations
- **Share Angular build artifact with amd64** — arm64 already downloads pre-built Angular, but amd64 rebuilds inline. Reuse the artifact for both, saving ~90s on release builds
- **Simplify build matrix** — Drop from 9 jobs to 5 on tag push: `unit-test → angular-build → build (alpine-amd64 + alpine-arm64) → publish → release`

### Image Optimizations
- **Standardize HEALTHCHECK to `wget`** — Alpine already uses `wget`; Debian uses `python urllib` with ~0.5s startup overhead. With Alpine-only, ensure HEALTHCHECK uses `wget`
- **Cache Python test image in Makefile** — Currently rebuilds every `make test` run. Add conditional build logic or use Docker layer cache to save ~30s per local test run

### Build Cache Optimization
- **Add BuildKit cache mount for npm** — Use `RUN --mount=type=cache,target=/root/.npm npm ci` so that on layer cache misses (dependency bumps, GHA cache eviction), npm reuses previously downloaded packages instead of re-fetching all ~150 packages. Saves ~30-60s on cache-miss rebuilds. Skip for pip (only 3 deps, negligible benefit).

## Migration

Users pulling `ghcr.io/nitrobass24/seedsync:latest` or version tags will automatically get the Alpine image. Document the change in CHANGELOG.md release notes.

### LINKED ISSUE #241: Accessibility: Convert all file action buttons to native <button> elements
## Summary

All six action buttons in the file action bar (`Queue`, `Stop`, `Extract`, `Validate`, `Delete Local`, `Delete Remote`) are currently rendered as `<div class="button">` elements with click-guard patterns and `[attr.disabled]` bindings. These are not keyboard accessible — keyboard users cannot activate them with Enter/Space, and screen readers do not announce them as interactive controls.

## Problem

Divs used as interactive controls lack:
- Native keyboard focus and activation (Enter/Space)
- Implicit ARIA `role="button"` semantics
- Correct `disabled` state signalling to assistive technologies

## Proposed Fix

In a single, uniform PR, convert **all** action buttons in `src/angular/src/app/pages/files/file.component.html` from:

```html
<div class="button" appClickStopPropagation
     [attr.disabled]="isXxx() ? null : true"
     (click)="!isXxx() || onXxx(file())">
```

to native `<button>` elements:

```html
<button type="button" class="button" appClickStopPropagation
        [disabled]="!isXxx()"
        (click)="onXxx(file())">
```

This must be done for **all** six buttons uniformly to maintain consistency across the action bar.

## Affected File

- `src/angular/src/app/pages/files/file.component.html`

## References

- Raised during PR review: https://github.com/nitrobass24/seedsync/pull/236#discussion_r2918762759
- PR: https://github.com/nitrobass24/seedsync/pull/236
- Requested by: @nitrobass24

### LINKED ISSUE #242: Add lftp xfer:verify toggle for inline transfer verification
## Summary

Add a UI toggle to enable lftp's built-in `xfer:verify` feature, which verifies file integrity inline during transfer by comparing checksums on source and destination immediately after each file completes.

This complements the existing post-download integrity check (#125 / PR #236) by catching corruption at transfer time before files enter the pipeline (extraction, staging moves, etc.).

## Motivation

- `xfer:verify` is zero custom code — lftp handles everything
- Catches corruption before extraction/staging runs, not after
- Per-file granularity — on failure, lftp retries just the failed file (up to `net:max-retries`), not the entire transfer
- Combined with the existing manual/auto-validate feature, gives users a layered integrity strategy

## Proposed UI

Add to the existing **Integrity Check** settings section:

1. **"Verify transfers inline (recommended)"** — checkbox, **default: enabled**
   - Maps to lftp settings `set xfer:verify true` and `set xfer:verify-command sha256sum`
   - When enabled, lftp automatically hashes each file after transfer and compares with remote

2. Keep the existing **"Auto-validate after download"** checkbox — **default: disabled**
   - This is the post-download SSH-based re-verification from PR #236
   - Acts as "belt and suspenders" for users who want a second check

## Implementation

- In `lftp.py`, add two settings when the toggle is enabled:
  ```
  set xfer:verify true
  set xfer:verify-command sha256sum
  ```
- Add the config property to `Config.Validate` (e.g., `xfer_verify`)
- Add the UI checkbox to `OPTIONS_CONTEXT_VALIDATE` in `options-list.ts`
- Wire through the settings page like existing options

## Notes

- If both `xfer:verify` and auto-validate are enabled, the staging deferral in auto-validate is redundant since lftp already verified the file. Consider simplifying by skipping deferral when `xfer:verify` is on — but this is optional and can be a follow-up.
- Need to research lftp's behavior when `xfer:verify` fails: does it retry automatically? Does it surface the error in a way SeedSync's log parser can detect?
- The `xfer:verify-command` should match the algorithm selected in the Integrity Check settings (md5sum/sha1sum/sha256sum).

## Context

Discussion from PR #236 review. See also how [RapidCopy fork](https://github.com/rccypher/RapidCopy) handles chunk-based validation for comparison.

## Milestone

v0.14.0

### LINKED ISSUE #265: File descriptor leak on restart causes OSError: [Errno 24]
## Description

When SeedSync restarts (e.g., after a config change via the web UI), the app crashes with `OSError: [Errno 24] No file descriptors available` during re-initialization.

The error occurs when creating multiprocessing primitives (Event/Semaphore) in `ExtractProcess.__init__()`, indicating file descriptors from the previous run are not being properly closed before restart.

## Stack Trace

```
File "/app/python/seedsync.py", line 131, in run
    controller = Controller(self.context, self.controller_persist)
File "/app/python/controller/controller.py", line 242, in __init__
    self.__extract_process = ExtractProcess()
File "/app/python/controller/extract/extract_process.py", line 66, in __init__
    super().__init__(name=self.__class__.__name__)
File "/app/python/common/app_process.py", line 53, in __init__
    self._terminate = Event()
...
OSError: [Errno 24] No file descriptors available
```

## Steps to Reproduce

1. Start SeedSync
2. Change a config setting via the web UI (triggers restart)
3. App crashes on restart with the above error
4. Requires a full container restart to recover

## Notes

- This is a regression of #234
- The LFTP pexpect process and SSH connections likely leak file descriptors across restarts
- A full container restart (`docker restart`) recovers the app since the OS reclaims all FDs

## Environment

- Platform: Unraid (Docker)
- SeedSync version: v0.13.4

### LINKED ISSUE #266: Expose General settings (debug, verbose, exclude_patterns) in web UI
## Description

The `[General]` config section (debug, verbose, exclude_patterns) is not exposed in the web UI settings page. Users must manually edit the config file to change these settings, which is problematic because:

1. **Config file edits get overwritten** — SeedSync periodically persists its in-memory config to disk (via `persist()`), so manual edits to the config file while the app is running are silently lost.
2. **Stop/edit/start is fragile** — Even editing while stopped can be overwritten if the timing is wrong or the config fails to load (falling back to defaults).

## Affected Settings

| Setting | Type | Purpose |
|---------|------|---------|
| `debug` | bool | Enable debug-level logging |
| `verbose` | bool | Enable verbose LFTP command logging (requires debug=True) |
| `exclude_patterns` | string | Comma-separated glob patterns to exclude from downloads |

## Expected Behavior

These settings should be viewable and editable from the Settings page in the web UI, just like LFTP, Controller, and AutoQueue settings are today.

## Notes

- `exclude_patterns` was added in v0.13.4 but is only configurable via the config file
- The verbose logging issue was discovered while debugging exclude patterns — the setting kept getting reset on restart because there was no way to set it through the UI

### LINKED ISSUE #267: LftpModel repeatedly re-adds same files every scan cycle
## Description

The LftpModel continuously logs "Adding file" for the same files on every scan cycle (~1 second intervals), even when nothing has changed. This creates excessive log noise and suggests the model is not correctly detecting that files already exist.

## Example Logs

```
22:53:35,344 - DEBUG - LftpModel: Adding file 'Boston.Blue.S01E11...FLUX.mkv [06e46a02]'
22:53:35,344 - DEBUG - LftpModel: Adding file 'Sherlock.Jr.1924...BHDStudio.mp4 [06e46a02]'
22:53:35,344 - DEBUG - LftpModel: Adding file 'The.Night.Agent.S03E08...ETHEL [06e46a02]'
...
(same files repeated every ~600ms)
```

Files that are actively downloading correctly show "Updating file" instead, but idle/queued files get re-added every cycle rather than being recognized as already present.

## Expected Behavior

Files should be added once, then subsequent scans should either skip them (no log) or log "Updating file" only if their state has changed.

## Impact

- Excessive debug log volume (hundreds of duplicate lines per minute)
- Potential unnecessary object churn in the model
- Makes it harder to find meaningful log entries

### LINKED ISSUE #271: Deleting stopped download leaves persist state stuck, ActiveScanner polls forever
## Bug Description

When deleting local files for a stopped download, the UI shows "Deleting" spinner forever and the ActiveScanner logs "Path does not exist" every second indefinitely. The persist state never updates.

## Steps to Reproduce

1. Start downloading a file (LFTP job running)
2. Stop the download
3. Delete local files via the UI
4. Observe: spinner never clears, ActiveScanner polls the (now deleted) path every ~1 second

## Root Cause

When an LFTP job disappears (whether completed or stopped), the file is added to both `downloaded_file_names` (persist) and `pending_completion` (in-memory). `pending_completion` keeps the ActiveScanner scanning the file until it reaches a terminal state.

The `pending_completion` exit conditions check for states: `DOWNLOADED`, `EXTRACTED`, `EXTRACT_FAILED`, `VALIDATED`, `CORRUPT`, `DELETED`. However, when `auto_delete_remote` is ON and the remote file still exists, the model builder's persist authority is **skipped** (to allow re-download). This means a locally-deleted file stays in `DEFAULT` state instead of transitioning to `DELETED`.

Since `DEFAULT` is not in the exit conditions, `pending_completion` is never cleared, and the ActiveScanner keeps polling the non-existent path forever.

## Fix

Add an early exit from `pending_completion` when a file is in `DEFAULT` state with `local_size is None` — meaning the local files are gone and there's nothing left to track.

## Logs

```
ActiveScanner.ActiveScanner - Path does not exist: /data/lftpsync/.../The.Night.Agent.S03E08...
```
(repeats every ~1 second indefinitely)

### PR DESCRIPTION
**Pull request by nitrobass24**
## Summary

- **Alpine-only Docker image** — Removed Debian variant; all images are now Alpine-based (~45 MB)
- **Integrity checking** — Inline transfer verification (xfer:verify) and optional post-download validation with per-file Validate button
- **Verbose logging in web UI** — Exposed verbose LFTP logging setting in the Settings page
- **CI deduplication** — Eliminated redundant amd64 Docker build on publish triggers
- **Stopped download delete stuck** — Fixed UI spinner and ActiveScanner polling forever after deleting local files for a stopped download
- **File descriptor leak on restart** — Multiprocessing Event/Queue references now released in `close_queues()`
- **Model rebuild log noise** — Temporary model objects no longer emit "Adding file" log messages every loop iteration
- **Exclude patterns** — Changed `--exclude` (regex) to `--exclude-glob` (glob) so patterns like `*.nfo` work correctly
- **Integrity check UX** — Renamed and reorganized validation settings for clarity; validate button shows tooltip when remote file is unavailable

## PRs Included

- #244 Alpine-only consolidation
- #265 FD leak fix
- #266 Verbose logging UI
- #267 Model log noise
- #268 FD leak terminate guard
- #269 Dummy logger propagation fix
- #270 General settings UI fix
- #271 Exclude-glob fix
- #272 Pending completion stuck fix
- #273 Active files cache invalidation fix
- #274 CI deduplication
- #276 Integrity check settings clarity and validate button UX

## Closes

Closes #125, Closes #156, Closes #225, Closes #228, Closes #231, Closes #241, Closes #242, Closes #265, Closes #266, Closes #267, Closes #271

## Test plan

- [ ] CI passes on this PR
- [ ] Docker image builds and starts successfully
- [ ] Web UI loads and settings page works
- [ ] Verify stopped download can be deleted without stuck spinner
- [ ] Integrity check settings render correctly
- [ ] Validate button shows tooltip when remote file unavailable

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Verbose LFTP logging option in Settings > Logging
  * Manual "Validate" button, validation statuses, and post-download validation mode

* **Improvements**
  * Docker image consolidated to Alpine-only (~45 MB), multi-arch (amd64/arm64)
  * Reworked integrity-check settings and UI wording
  * CI pipeline: finer-grained build/test/push steps and improved container readiness checks

* **Bug Fixes**
  * Fixed stuck downloads on delete, FD exhaustion on restart, noisy rebuild logs, and LFTP exclude pattern handling

* **Version**
  * Updated to v0.14.0

### CODE REVIEW

File: website/docs/configuration.md [hunk 1]
```diff
@@ -58,6 +58,30 @@ Configure exclude patterns per path pair, or in the main settings when using a s
...
+
+### Transfer verification (inline)
+
+- **Verify transfers inline (recommended)**: When enabled, LFTP compares checksums during the download itself, catching corruption in real-time. Enabled by default.
+
+### Post-download validation
+
+- **Enable post-download validation**: Enables a separate validation step after download completes, comparing local and remote checksums via SSH. Disabled by default.
+- **Auto-validate after download**: Automatically queue files for validation when their download finishes. Requires post-download validation to be enabled.
+- **Hash Algorithm**: Algorithm used for both inline verification and post-download validation. Options: `md5`, `sha1`, `sha256`.
```
[the remark this answers is not in the document]
Comment (nitrobass24) [reply website/docs/configuration.md]: Hash Algorithm scope: Not confirmed. At controller.py:432: lftp.xfer_verify_command = "{}sum".format(validate_cfg.algorithm) — the algorithm IS used
  for both inline xfer:verify AND post-download validation. My docs saying "both" were actually correct. The UI description in options-list.ts is the incomplete
  one. No docs change needed
